### ISSUE TITLE
How to setup a toolchain file

### ISSUE THREAD

**Issue by pfultz2** [mesonbuild/meson#2472]
I am wanting to invoke meson using the same toolchain setting that are setup by my cmake toolchain. I would like to automatically generate a file(or meson command) to do this. I see there is `--cross-file` to pass in a file with settings, but does this work for native builds as well? Also, is there documentation on what settings are in there?

Also, I can set `--prefix` to set installation prefix, but how do I set the prefix for searching for dependencies(like cmake's `CMAKE_PREFIX_PATH`)?

**Comment by jpakkane** [comment]
For native builds we follow the Unix established practice for this, meaning that the specifying the `CC`, `CXX` etc environment variables when the build directory is first generated selects that compiler. If they are not set then we will autodetect a compiler and use that. This is arguably not as smooth on Windows, but usually you have only one compiler in your path so it will be chosen automatically.

> how do I set the prefix for searching for dependencies(like cmake's CMAKE_PREFIX_PATH)

We don't have a concept like that currently. On any platform that is not Windows you should not do library lookups manually whenever you can avoid it but use pkg-config files instead. Then it is a question of setting `PKG_CONFIG_PATH` envvar to your desired value. Pkg-config also works on Windows but not many people use it there yet, I guess.

If you provide dependencies by building them as subprojects or use prebuilt libraries provided via wrap files, this issue does not come up since the deps are not searched for in the system.

**Comment by pfultz2** [comment]
> For native builds we follow the Unix established practice for this, meaning that the specifying the CC, CXX etc environment variables when the build directory is first generated selects that compiler.

Ok I see. Usually autoconf scripts show this under `--help`. You should add that to meson's help as well. 

Also, does this mean that `CFLAGS`, `CXXFLAGS`, and `LDFLAGS` can be used?

> Then it is a question of setting PKG_CONFIG_PATH envvar to your desired value.

I see, easy enough.

> Pkg-config also works on Windows but not many people use it there yet, I guess.

So how does meson find the dependencies on windows?

> If you provide dependencies by building them as subprojects or use prebuilt libraries provided via wrap files, this issue does not come up since the deps are not searched for in the system.

I am actually wanting to invoke meson from cget. So cget will already install the dependencies, so its matter of telling meson where they are.

**Comment by liugang** [comment]
> I am actually wanting to invoke meson from cget. So cget will already install the dependencies, so its matter of telling meson where they are.

you can use `add_project_arguments` and `add_project_link_arguments`:
example:
```
compile_args += ['-Ipath/to/include']
link_args += ['-Lpath/to/lib', '-lfoo', '-lbar']
add_project_arguments(libahal_compile_args, language: 'c')
add_project_link_arguments(link_args, language: 'c')
```

**Comment by pfultz2** [comment]
> you can use add_project_arguments and add_project_link_arguments:

But doesn't that require modifying the build scripts? I would like to avoid modifying files.

**Comment by liugang** [comment]
Are you using third-party package which using mesonbuild system?
if so how it search the dependency, pkg-config? find_library() ?

**Comment by pfultz2** [comment]
> Are you using third-party package which using mesonbuild system?

No, not necessarily. 

I am wanting to setup cget to build meson packages. I have a description of a toolchain, which can include the compiler, flags to the compiler, compile options, search paths for dependencies, etc. I would like to invoke meson with those same settings. I can use the environment variables to set the compiler, and pkgconfig search paths, but how do I set the other toolchain flags such as compiler flags?

It looks like I can set all of this in the `--cross-file` but I am not sure I can use that for native builds. It would be nice if there was a toolchain file that could be created similar to the crossfile, but can be used for native builds as well.

**Comment by pfultz2** [comment]
So it does look like it reads the other environment variables for flags here: https://github.com/mesonbuild/meson/blob/7c2ae55fab2f0837531188373d276f168845e62f/mesonbuild/environment.py#L71

**Comment by jpakkane** [comment]
Yes, we do support all the standard Unix style envvars, but note that they are only added when first running Meson for a new build dir. After that their values are ignored, but you can change these settings, like all others, with `meson configure`.

**Comment by pfultz2** [comment]
> but note that they are only added when first running Meson for a new build dir.

Thats perfect.

> After that their values are ignored, but you can change these settings, like all others, with meson configure

Interesting.

**Comment by pfultz2** [comment]
So `PKG_CONFIG_PATH` controls where pkg-config searches for a library, but how do I control where `find_library` searches for libraries?

**Comment by liugang** [comment]
You can set environment `LDFLAGS`

```
cc = meson.get_compiler('c')
zlib = cc.find_library('z', require: false)

if zlib.found()
    message('zlib library found')
else
    message('zlib library not found')
endif
```

then set LDFLAGS to meson
```
LDFLAGS=-L/path/to/library meson builddir
```

you can check builddir/meson-logs/meson-log.txt to see how it works

**Comment by jpakkane** [comment]
> but how do I control where find_library searches for libraries

There is no global option for that currently.

**Comment by cenit** [comment]
>> but how do I control where find_library searches for libraries

> There is no global option for that currently.

any news on this front? In order to integrate better meson with vcpkg I'd need that...

**Comment by pfultz2** [comment]
Maybe there should be a separate issue opened for adding a flag to meson to tell it where to search for dependencies.

**Comment by jpakkane** [comment]
Note that by default Meson uses the compiler to look up libraries and also extracts that list for some lookups. Thus if your compiler is set up to look in certain directories, it should work out of the box.

I'm not sure how MSVC works, but I _think_ that this is not really possible, at least not in the same way. I'm also not sure, but I seem to recall that if you add the equivalent of `-L` to your link flags in the cross file, they are passed to the library lookup code that passes them to the compiler when running the search.

**Comment by cenit** [comment]
Luckily meson listen to a `CMAKE_PREFIX_PATH` environment variable (which is wrong from the CMake point of view, but whatever)
Using this trick, I have been able to at least start my work and integrate meson with the rest of the libraries that export a <portfile>Config.cmake file

**Comment by cenit** [comment]
> by default Meson uses the compiler to look up libraries and also extracts that list for some lookups. Thus if your compiler is set up to look in certain directories, it should work out of the box.

I just realized that it might be possible that allowing the vcpkg-msbuild integration maybe some libraries will come for free? I have to check for that, thanks for the (hopeful) gem that I was overlooking

**Comment by pfultz2** [comment]
> Note that by default Meson uses the compiler to look up libraries and also extracts that list for some lookups. Thus if your compiler is set up to look in certain directories, it should work out of the box.

So adding `-L` and `-I` flags works for find_library and for finding boost? Its not as nice(or easy to use) as cmake's `CMAKE_PREFIX_PATH`, but I guess it works.

**Comment by bonzini** [comment]
Spring cleaning of support/RFC issues; mostly obsolete due to introduction of native files.
